Security hardening: Expose Build flags for Position Independed Execution (PIE)

crlorentzen opened the pull request:
Security hardening: Expose Build flags for Position Independed Execution (PIE)

Adds build flags to allow enabling support for ASLR in Openssl binaries via the Configure process.

~~CLA: trivial~~

By default Openssl does have the option to enforced the PIE hardening standard to support ASLR for binaries such as `openssl`

* Testing done on Amazon Linux 2023
```
sudo yum groupinstall "Development Tools"

git clone https://github.com/openssl/openssl.git

cd openssl
./Configure --strict-warnings enable-pie
make

# After the change PIE enabled (LSB pie executable)
file apps/openssl
apps/openssl: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=17ec32ebd630804e5dcfee3ef5dba72ece707334, for GNU/Linux 3.2.0, not stripped

# Confirm with no flag we maintain no-PIE (LSB executable)
./Configure --strict-warnings 
make clean
make
file apps/openssl
apps/openssl: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=11f5f229eec64a45e9f63da88a3065368eba6227, for GNU/Linux 3.2.0, not stripped
```

Some OS distributions enable PIE by default, e.g. Ubuntu https://wiki.ubuntu.com/ToolChain/CompilerFlags, and the OS distribution of openssl is generally built with PIE enabled, however, the default from OpenSSL does not implement the PIE hardening feature enabling ASLR functionality for the openssl binaries.

t8m commented:
Please do not add merge commits.

t8m commented:
I am not  sure about this - we aim to support fairly old or special systems, are we sure it won't break on these?

crlorentzen commented:
My apologies I did not intend to merge the master commits in this PR. Please let me know what I can and should do to repair.

crlorentzen commented:
> I am not sure about this - we aim to support fairly old or special systems, are we sure it won't break on these?

I'll need to rely on The OpenSSL project to decide on this question.

t8m commented:
In your case `git reset --hard HEAD~ && git push --force` should repair it.

t8m commented:
Hmm... you can even see this CI failure which is related.

hlandau commented:
We can see that this is causing some issues on m68k.

I think this is a good idea and we should do this. We may need to exempt a few older architectures and platforms, but I don't see a big issue with doing that.

The fuzzer CI failure is a bit odd:
```
2023-11-02T15:46:47.9142079Z /usr/bin/ld: final link failed: No space left on device
2023-11-02T15:46:47.9367251Z clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
```

paulidale commented:
I don't think we should make this mandatory: old platforms, folks who don't want it, etc.

My suggestion would be for this to be a separate configuration target or done as a configuration option (`enable-pie` or `disable-pie`) or documented in the INSTALL.md (or similar).  I.e. make it available but not mandatory.

hlandau commented:
> I don't think we should make this mandatory: old platforms, folks who don't want it, etc.
> 
> My suggestion would be for this to be a separate configuration target or done as a configuration option (`enable-pie` or `disable-pie`) or documented in the INSTALL.md (or similar). I.e. make it available but not mandatory.

I agree.

I think it should definitely be enabled by default on primary platforms.

t8m commented:
**OTC: This should be a build-time option. Which can be on by default for selected platforms after they are tested to work.**

slontis commented:
@crlorentzen
If you want this PR to be accepted you will have to add it as a configurable build option
Refer to INSTALL.md 'Enable and Disable Features' for examples of existing ones.
You would need to edit configdata.pm.

crlorentzen commented:
Looking for configdata.pm I find configdata.pm.in and I am not sure what is needed there.

Requesting feedback now as I cannot devote time to this at the moment, but don't want to lose information if I can return to this in the futue. Plus, maybe someone else is also interested in understanding the process.

I see the Configure script will require the addition of `pie` as a [disableable](https://github.com/openssl/openssl/blob/9eabb30ab4491bdcf49c5bfeef659ca846da5160/Configure#L414) and then a default [disabled](https://github.com/openssl/openssl/blob/9eabb30ab4491bdcf49c5bfeef659ca846da5160/Configure#L563)

after those changes one would add to [00-base-templates.conf](https://github.com/openssl/openssl/blob/9eabb30ab4491bdcf49c5bfeef659ca846da5160/Configurations/00-base-templates.conf#L72) something like
```
...
bin_cflags       => 
  sub {
    my @flags = ();
    if (!defined($disabled{pie}) {
      push(@flags, "-fPIE");
    }
    return (" ", @flags);
  },
bin_lflags       => 
  sub {
    my @flags = ();
    if (!defined($disabled{pie}) {
      push(@flags, "-pie");
    }
    return (" ", @flags);
  },
...
```

Then remove the changes made to 10-base-templates.conf.

Does the above sound correct?

slontis commented:
I mis-remembered how you change it..

You edit
Configure +
install.md

You have the right idea for the conf file.
The decision as to whether the option is on by default may be up for debate..

crlorentzen commented:
This Pull request is now a commit to add the `enable-pie` flag suggested by the OTC

t8m commented:
We will need a CLA to be signed as we cannot accept this contribution with CLA: trivial.

As this is disabled by default there should be a CI run in runchecker-merge.yml added that enables this.

crlorentzen commented:
@t8m Understood about CLA: trivial. Will look at signing and providing that document.

Also took a stab at the `run-checker-merge.yml` requirement

crlorentzen commented:
Closing and Re-openning per guidance to recheck CLA database

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

crlorentzen commented:
Bump; Requesting OTC review and feedback on how to proceed getting this change to allow PIE as part of the Openssl Configuration process.

t8m commented on INSTALL.md:
```diff
@@ -861,6 +861,10 @@ As synonym for `no-padlockeng`.  Deprecated and should not be used.
 
+### enable-pie
+
+Build with suport for Position Independent Execution.
```
Typo: support

t8m reviewed (changes requested):
Could you please add a CHANGES.md entry?

crlorentzen commented:
Fixed typo and added CHANGES.md entry.

crlorentzen commented:
I think I successfully fixed the CHANGES.md conflict

t8m commented:
Unfortunately there are still conflicts in CHANGES.md and there is a merge commit which is not allowed.

crlorentzen commented:
eventually I may be able to get back to trying to make this acceptable

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

t8m reviewed (approved):
LGTM but please rebase again to fix the conflict in CHANGES.md

crlorentzen commented:
Thank you @t8m, rebased and fixed conflicts in CHANGES.md. Moving the change information to "Changes between 3.3 and 3.4 [xx XXX xxxx]" since it didn't make for 3.3.

tom-cosgrove-arm reviewed (approved):
LGTM, subject to CI

tom-cosgrove-arm commented:
@t8m can you reconfirm?

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been  updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

t8m commented:
Squashed and merged to the master branch. Thank you for your contribution.
